sched/distributor/poolDistributor_test.go

```diff
@@ -1,230 +1,127 @@
-package distributor
+package distributor_test
 
 import (
-	"github.com/golang/mock/gomock"
-	cm "github.com/scootdev/scoot/sched/clustermembership"
-	"strings"
+	"github.com/scootdev/scoot/cloud/cluster"
+	"github.com/scootdev/scoot/cloud/cluster/memory"
+	"github.com/scootdev/scoot/sched/distributor"
 	"testing"
 )
 
-func TestPoolDistributor_NilCluster(t *testing.T) {
-	mockCtrl := gomock.NewController(t)
-	defer mockCtrl.Finish()
-
-	clusterMock := cm.NewMockCluster(mockCtrl)
-	clusterMock.EXPECT().Members().Return(nil)
-
-	dist := NewPoolDistributor(clusterMock)
-
-	if nil != dist {
-		t.Error("Expected dist to be nil when no members in Cluster")
-	}
-}
-
-func TestPoolDistributor_EmptyCluster(t *testing.T) {
-	mockCtrl := gomock.NewController(t)
-	defer mockCtrl.Finish()
-
-	nodes := make([]cm.Node, 0)
-
-	clusterMock := cm.NewMockCluster(mockCtrl)
-	clusterMock.EXPECT().Members().Return(nodes)
-
-	dist := NewPoolDistributor(clusterMock)
-
-	if nil != dist {
-		t.Error("Expected dist to be nil when no members in Cluster")
-	}
-}
-
 func TestPoolDistributor_OneNodeCluster(t *testing.T) {
-	mockCtrl := gomock.NewController(t)
-	defer mockCtrl.Finish()
-
-	nodes := cm.GenerateTestNodes(1)
-	clusterMock := cm.NewMockCluster(mockCtrl)
-	clusterMock.EXPECT().Members().Return(nodes)
-
-	dist := NewPoolDistributor(clusterMock)
-
-	if nil == dist {
-		t.Error("Expected dist to not be nil when 1 or more nodes in Cluster")
-	}
-
-	node1 := dist.ReserveNode()
-
-	select {
-	case <-dist.freeCh:
-		t.Error("Expected cluster to be full schedlued")
-	default:
-	}
-
-	dist.ReleaseNode(node1)
-	node2 := dist.ReserveNode()
-
-	if node1.Id() != node2.Id() {
-		t.Error("Expected nodes to be the same Id since there is only one in the cluster")
-	}
-
-	select {
-	case <-dist.freeCh:
-		t.Error("Expected cluster to be full schedlued")
-	default:
-	}
+	h := makeHelper(t, "node1")
```

Comment:
Yes, mock-based tests can be changed to this style.

I think that mock tests are somewhat harder to read, but that's not the reason they're bad.

They're bad because they're brittle. Take our current structure: Pool is a Library that's used by Scheduler. If we refactor Pool, of course we have to change Pool's tests. And we have to change Scheduler. But if Scheduler's tests use mock Pools instead of fake Pools, then we also have to refactor Scheduler's tests, and these changes are often laborious.

Mock-based tests discourage refactors. Ooof. That's a big one.

Mock-based tests are the right test when you want to test exactly how Scheduler calls Pool, which often happens when you need to meet certain performance goals (e.g., make sure that we're making batch requests). But in most cases, you want to test that Scheduler gets its job done. Mock-based tests also test how Scheduler does its job. That's more work to test the wrong thing.